### ISSUE TITLE
Error parsing a simple dot file

### ISSUE THREAD

**Issue by dilawar** [ivan-m/graphviz#3]
Here is the code I am running.

~~~~
import Data.GraphViz
import Data.GraphViz.Commands.IO
import Data.GraphViz.Types
import Data.Graph
import System.Environment
import qualified Data.Text as T

read_dot_file :: String -> IO (DotGraph String)
read_dot_file filename = readDotFile filename

main = do
    files <- getArgs
    dots <- mapM read_dot_file files
    print $ dots
    putStrLn "Done"
~~~~

And the following file is passed as argument.

     $ runghc channel.hs channel.dot

~~~~
digraph channel {
    x0 -> y1, y2, y3 [ label = "1/3" ];
    x1 -> y2, y3, y4 [ label = "1/3"];
    x2 -> y3, y4, y5 [ label = "1/3"];
    x3 -> y4, y5, y6 [ label = "1/3" ];
    x4 -> y5, y6, y7 [ label = "1/3" ];
    x5 -> y6, y7, y8 [ label = "1/3" ];
    x6 -> y7, y8, y9 [ label = "1/3" ];
    x7 -> y8, y9, y10 [ label = "1/3" ];
    x8 -> y9, y10, y0 [ label = "1/3" ];
    x9 -> y10, y0, y1 [ label = "1/3" ];
    x10 -> y0, y1, y2 [ label = "1/3" ];
}
~~~~

I am getting the following error:

~~~~
channel.hs: Error when parsing Dot code:
Unable to parse the Dot graph; usually this is because of either:
  * Wrong choice of representation: try the Generalised one
  * Wrong choice of node type; try with `DotGraph String`.

The actual parsing error was:
	Not a valid value wrapped in braces.
	Missing closing bracket:
	Not the expected character: }

shell returned 1
~~~~

What am I doing wrong?

**Comment by dilawar** [comment]
ghc version 7.6.3 and graphviz-2999.18.0.1.

**Comment by dilawar** [comment]
Works well with following dot content:

~~~~
digraph simpleChannel {
    xA -> yA;
    xB -> yB;
    xC -> yC;
    xD -> yD [ label = "8/9" ];
    xD -> yA [ label = "1/9" ];
}
~~~

**Comment by dilawar** [comment]
I see, statement line `xA -> yA` are supported but not `xA -> yA, yB`.

This should be a feature request then.

**Comment by ivan-m** [comment]
*sigh* yet another implicit feature that Graphviz accepts that isn't specified in the Dot language.

I'll work on this and try and get it going as soon as I can.

**Comment by ivan-m** [comment]
Specifically, the [language definition](http://graphviz.org/content/dot-language) states that:

...
*edge_stmt* 	: 	(*node_id* | *subgraph*) *edgeRHS* [ *attr_list* ]
...
*subgraph* 	: 	[ **subgraph** [ *ID* ] ] **'{'** *stmt_list* **'}'**
...

So it really *should* have braces there.

**Comment by ivan-m** [comment]
Please let me know if this fixes it for you.

I'm going to wait before applying this to master until I've heard a response to my query about this to the [mailing list](http://graphviz.996277.n3.nabble.com/Edge-statements-in-Dot-td4255.html).

**Comment by dilawar** [comment]
Yes. It does. Thank you!

In my opinion, you can have two function to parse graphviz file: one is strict which honours the official grammar, and second is more relaxed like `dot` `neato` etc.

**Comment by ivan-m** [comment]
I'm re-opening this so I can keep track of it.

I am planning on a large re-structure, but don't want to do it just yet (ask me for anything except time... ;-)

**Comment by ivan-m** [comment]
OK, looks like this is the new intended behaviour of `dot`, etc. so I'll merge this in.

**Comment by jwiegley** [comment]
It would be nice to get a clearer idea of what's actually wrong with the file.  I'm getting the exact same error as above, although I'm pretty sure there's not a problem at the very beginning of the file, as the error indicates.

**Comment by ivan-m** [comment]
@jwiegley do you have a sample file I can look at?

**Comment by jwiegley** [comment]
So, I found my problem eventually.  I was trying to parse the following as a `DotGraph Int`:

```
digraph { a -> b }
```

The error I got was:

```
The actual parsing error was:
    Not a valid value wrapped in braces.
    Missing closing bracket:
    Not the expected character: }
```

What I would have expected was something more like:

```
Read error at: 'a'
```

Of course it was easily fixed by switching to `DotGraph Text`, but it took me a while to realize that this was the problem, since it's my first time using the graphviz parser.  The fact that the problem had nothing really to do with brackets confused me.
